Sentinel

sjpotter opened the pull request:
### Description

<!-- Please provide a description of the change below, e.g What was the purpose? -->
<!-- Why does it matter to you? What problem are you trying to solve? -->
<!-- Tag in any linked issues. -->

> Describe your pull request here

---

### Checklist

<!-- Please make sure to review and check all of these items: -->

- [ ] Does `npm test` pass with this change (including linting)?
- [ ] Is the new or changed code fully tested?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?

<!-- NOTE: these things are not required to open a PR and can be done
afterwards / while the PR is open. -->

sjpotter committed:
redis client socket changes needed for sentinel

sjpotter committed:
add pooling

sjpotter committed:
Sentinel Implementation [EXPERIMENTAL]

sjpotter commented:
note, tsconfig.base.json as is, doesn't transpile import into request, which gives m problems when testing by hand.

chayim commented on packages/client/lib/sentinel/utils.ts:
```diff
@@ -0,0 +1,25 @@
+  return { "host": data[3], "port": Number(data[5]) };
+}
+
+export function createNodeList(data: any) {
```
There's a lot of any for typing here...

sjpotter replied:
yes, I dont know how to get typing out of raw redis commands yet.

chayim commented on packages/client/lib/sentinel/utils.ts:
```diff
@@ -0,0 +1,25 @@
+
+  for (const nodeData of data) {
+    const node = parseNode(nodeData)
+    if (node === undefined) {
```
undefined is like false in node - can we do an if/else here?

sjpotter replied:
what do you mean?  in this case, I have parseNode return undefined for a node that is down, so can ignore it.

chayim commented on packages/client/lib/sentinel/pub-sub-proxy.ts:
```diff
@@ -0,0 +1,189 @@
+    return resp;
+  }
+
+  async pSubscribe<T extends boolean = false>(
```
is this pSubscribe/pUnsubscribe because of camelCase? I would have expected this to follow the redis command name (psubscribe/punsubscribe)

sjpotter replied:
this is copying from rest of node api I believe.

chayim commented on packages/client/lib/sentinel/pub-sub-proxy.ts:
```diff
@@ -0,0 +1,189 @@
+    return resp;
+  }
+
+  async pUnsubscribe<T extends boolean = false>(
```
same

chayim commented on packages/client/lib/sentinel/multi-commands.ts:
```diff
@@ -0,0 +1,215 @@
+);
+
+export default class RedisSentinelMultiCommand<REPLIES = []> {
+  private static _createCommand(command: Command, resp: RespVersions) {
```
a lot of boilerplating - can this be done better

sjpotter replied:
copied/pasted from cluster object

sjpotter committed:
Merge remote-tracking branch 'leibale/v5' into sentinel

sjpotter committed:
improve typing with SENTINEL_ client members

sjpotter committed:
cleanup - remove unused comments / commented code

sjpotter committed:
small sendCommand change + revert change to tsconfig

sjpotter committed:
add more sentinel commands needed for testing.

sjpotter committed:
lots of fixups and a reasonable first pass test suite

sjpotter committed:
add a timer option to update topology in background

+ don't need both sentinel client and pubsubclient
+ nits

sjpotter committed:
format all the things

sjpotter committed:
more progress

sjpotter committed:
Merge remote-tracking branch 'upstream/v5' into sentinel

sjpotter committed:
small cleanup

sjpotter committed:
try to group promises together to minimize the internal await points

sjpotter committed:
redo events, to keep a single topology event to listen on

sjpotter committed:
nits + readme

sjpotter committed:
add RedisSentinelFactory to provide lower level access to sentinel

sjpotter committed:
Merge remote-tracking branch 'upstream/v5' into sentinel

sjpotter committed:
nit

sjpotter committed:
update

sjpotter committed:
add RedisSentinelClient/Type for leased clients
	returned by aquire()
	used by function passed to use()

sjpotter committed:
add self for private access + improve emitting

sjpotter committed:
nit

sjpotter committed:
nits

sjpotter committed:
improve testing

- improve steady state waiting between tests
- get masternode from client, not from sentinels themselves (not consistent and then client isn't changing as we expect
- provide extensive logging/tracing on test errors
	- provide a very low impact tracing mechanism withinthe code that only really impacts code when tracing is in use.

sjpotter committed:
ismall nit for typing

sjpotter committed:
bunch of changes

- harden testing
	- don't use sentinel[0] for debug error dump as could be downed by a test
	- increase time for sentinel down test to 30s (caused a long taking failover)
- add client-error even / don't pass throuh client errors as errors option for pubsub proxy
- when passing through cient errors as error events, dont pass the event, but the Error object, as only Error objects are supposed to be on 'error'
	-

sjpotter committed:
improve pub sub proxy.

save the refference to all channel/pattern listeners up front on creation, dont hve to fetch the object each time, as it doesn't change.

removes race condition between setting up the listener and the pub sub node going down and being recreated.

sjpotter committed:
wrap the passed through RedisClient error to make clear where its coming from.

sjpotter committed:
refactor sentinel object / factory tests apart

sjpotter committed:
harden tests a little bit more

sjpotter committed:
add pipeline test

sjpotter committed:
add scripts/function tests + fixups / cleanups to get them to work

sjpotter committed:
change to use redis-stack-server for redis nodes to enable module testing

sjpotter committed:
fix test, forgot to return in use function with module

sjpotter committed:
rename test

sjpotter committed:
improve tests to test with redis/sentinel nodes with and withput passwords

this tests that we are handling the nodeClientOptions and sentinelClientOptions correctly

sjpotter committed:
cleanup for RedisSentinel type generic typing in tests

sjpotter committed:
remove debugLog, just rely on traace mechanism

sjpotter committed:
added multi tests for script/function/modules

sjpotter committed:
don't emit errors on lease object, only on main object

sjpotter committed:
improve testing

sjpotter committed:
extract out common code to reduce duplication

sjpotter committed:
nit

sjpotter committed:
nits

sjpotter committed:
nit

sjpotter committed:
remove SENTINEL_... commands from main client, load them via module interface

sjpotter committed:
missed adding RedisSentinelModule to correct places in RedisSentinelFactory

sjpotter committed:
nits

sjpotter committed:
fix test logging on error

1) it takes a lot of time now, so needs larger timeout
2) docker logs can be large, so need to increase maxBuffer size so doesn't error (and break test clean up)

sjpotter committed:
invalidate watches when client reconnects

+ provide API for other wrapper clients to also create invalid watch states programatically.

Reasoning: if a user does a WATCH and then the client reconnects, the watch is no longer active, but if a user does a MULTI/EXEC after that, they wont know, and since the WATCH is no longer active, the request has no protection.

The API is needed for when a wrapper client (say sentinel, cluster) might close the underlying client and reopen a new one transparently to the user.  Just like in the reconnection case, this should result in an error, but its up to the wrapping client to provide the appropriate error

leibale committed:
remove WATCH and UNWATCH command files, fix WATCH and UNWATCH return type, some more cleanups

leibale committed:
missing file in last commit :P

leibale committed:
support for custom message in `WatchError`

leibale committed:
setDirtyWatch

leibale committed:
update watch docs

sjpotter committed:
Merge branch 'watch-invalidate' into sentinel

sjpotter committed:
fixes needed

sjpotter committed:
Merge branch 'watch-invalidate' into sentinel

sjpotter committed:
wip

sjpotter committed:
get functions/modules to work again

self -> _self change

sjpotter committed:
reuse leased client on pipelined commands.

though I realize this implementation, really only works after the first write command.

unsure this is worth it.

sjpotter committed:
test tweaks

sjpotter committed:
nit

sjpotter committed:
change how "sentinel" object client works, allow it to be reserved

no more semaphore type counting

leibale committed:
Merge branch 'v5' of github.com:redis/node-redis into sentinel

leibale committed:
Merge branch 'v5' of github.com:redis/node-redis into sentinel

leibale committed:
review

leibale committed:
Merge branch 'v5' of github.com:redis/node-redis into sentinel

leibale committed:
Merge branch 'v5' of github.com:redis/node-redis into sentinel

sjpotter committed:
fixes to get more tests to pass

leibale committed:
Merge branch 'v5' of github.com:redis/node-redis into sentinel

leibale committed:
Merge branch 'v5' of github.com:redis/node-redis into sentinel

leibale committed:
handle dirtyWatch and watchEpoch in reset and resetIfDirty

sjpotter committed:
"fix", but not correct, needs more work

leibale committed:
fix pubsub proxy

sjpotter committed:
remove timeout from steadyState function in test, caused problems

sjpotter committed:
improve restarting nodes

sjpotter committed:
fix pubsub proxy and test
